packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+	);
+
+	return {
+		...otherProps,
+		className: classes
+	};
+}
+
+// in `component.tsx`
+function Example(
+	props: PolymorphicComponentProps< ExampleProps, 'div' >,
+	forwardedRef: Ref< any >
+) {
+	const exampleProps = useExampleComponent( props );
+
+	return <View { ...spacerProps } ref={ forwardedRef } />;
+}
+```
+
+A couple of good examples of how hooks are used for composition are:
+
+- the `Card` component, which builds on top of the `Surface` component by [calling the `useSurface` hook inside its own hook](/packages/components/src/card/card/hook.js);
+- the `HStack` component, which builds on top of the `Flex` component and [calls the `useFlex` hook inside its own hook](/packages/components/src/h-stack/hook.js).
+
+<!-- ### APIs Consinstency
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`), `has*` (e.g. `hasValue`) or `enable*` (e.g. `enableScroll`)
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- Subcomponents naming conventions (e.g `CardBody` instead of `Card.Body`)
+- ...
+
+### Performance
+
+TDB -->
+
+### Technical requirements for new components
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group) as an example.
+
+#### TypeScript
+
+We strongly encourage using TypeScript for all new components. Components should be typed using the `WordPressComponent` type (more details about polymorphism can be found above in the "Components composition" section).
+
+#### Emotion
+
+All new component should be styled using [Emotion](https://emotion.sh/docs/introduction).
+
+Note: Instead of using Emotion's standard `cx` function, the custom [`useCx` hook](/packages/components/src/utils/hooks/use-cx.ts) should be used instead.
+
+#### Context system
+
+The `@wordpress/components` context system is based on [React's `Context` API](https://reactjs.org/docs/context.html), and is a way for components to adapt to the "context" they're being rendered in.
+
+Components can use this system via a couple of functions:
+
+- they can provide values using a shared `ContextSystemProvider` component
+- they can connect to the Context via `contextConnect`
+- they can read the "computed" values from the context via `useContextSystem`
+
+An example of how this is used is the [`Card` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/card/card/component.js#L44-L54) — all of its subcomponents (e.g. `CardBody`, `CardHeader`...) will [use, by default, the same value for the `size` as the parent `Card` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/card/card-body/hook.js#L23) — which results in all the components having all the correct spacing "auto-magically".
```

Comment:
The link to the lines `Card` and `CardBody` components might quickly get out misleading when lines are added or removed. Can we copy this snippet here and link to the files instead?

Reply:
I've updated this section by adding a code snippet showing how the `CardBody` component "inherits" the `size` prop from the parent `Card` via the Context System in [`c34f731` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/c34f73168aafb25ab9412d3f0c4f5ed59c3c1358)